src/NuGet.Protocol.VisualStudio/PowerShellSearchResourceV2.cs

```diff
@@ -1,32 +1,126 @@
 using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
+using NuGet.Packaging.Core;
 using NuGet.Protocol.Core.Types;
+using NuGet.Protocol.Core.v2;
+using NuGet.Versioning;
 
 namespace NuGet.Protocol.VisualStudio
 {
     public class PowerShellSearchResourceV2 : PSSearchResource
     {
-        private readonly UISearchResource uiSearchResource;
+        private IPackageRepository V2Client { get; }
 
-        public PowerShellSearchResourceV2(UISearchResource search)
+        public PowerShellSearchResourceV2(V2Resource resource)
         {
-            uiSearchResource = search;
+            V2Client = resource.V2Client;
         }
 
-        public override async Task<IEnumerable<PSSearchMetadata>> Search(string search, SearchFilter filters, int skip, int take, CancellationToken token)
+        public override async Task<IEnumerable<PSSearchMetadata>> Search(string searchTerm,
+            SearchFilter filters,
+            int skip,
+            int take,
+            CancellationToken cancellationToken)
         {
-            var searchResults = await uiSearchResource.Search(search, filters, skip, take, token);
-            return searchResults.Select(item => GetPSSearch(item));
+            return await GetSearchResultsAsync(searchTerm, filters, skip, take, cancellationToken);
         }
 
-        private PSSearchMetadata GetPSSearch(UISearchMetadata item)
+        private async Task<IEnumerable<PSSearchMetadata>> GetSearchResultsAsync(string searchTerm,
+            SearchFilter filters,
+            int skip,
+            int take,
+            CancellationToken cancellationToken)
         {
-            return new PSSearchMetadata(item.Identity, item.Versions.Select(v => v.Version), item.Summary);
+            return await Task.Run(() =>
+            {
+                var query = V2Client.Search(
+                    searchTerm,
+                    filters.SupportedFrameworks,
+                    filters.IncludePrerelease);
+
+                // V2 sometimes requires that we also use an OData filter for 
+                // latest /latest prerelease version
+                if (filters.IncludePrerelease)
+                {
+                    query = query.Where(p => p.IsAbsoluteLatestVersion);
+                }
+                else
+                {
+                    query = query.Where(p => p.IsLatestVersion);
+                }
+
+                query = query
+                    .OrderByDescending(p => p.DownloadCount)
+                    .ThenBy(p => p.Id);
+
+                // Some V2 sources, e.g. NuGet.Server, local repository, the result contains all 
+                // versions of each package. So we need to group the result by Id.
+                var collapsedQuery = query.AsEnumerable().AsCollapsed();
+
+                // execute the query
+                var allPackages = collapsedQuery
+                    .Skip(skip)
+                    .Take(take)
+                    .ToList();
```

Comment:
Just curious. Is not .ToArray() better than .ToList()? If so, same comment for the code below